File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/motor/MotorControl.java
package org.firstinspires.ftc.teamcode.motor;


import androidx.annotation.NonNull;

import com.qualcomm.robotcore.hardware.*;

import org.firstinspires.ftc.robotcore.external.navigation.CurrentUnit;
import org.firstinspires.ftc.teamcode.helpers.control.PIDFController;

import java.util.ArrayList;

/**
 * This class is used to control the motor systems on the robot.
 */
public class MotorControl {

    public final Claw extendoClaw;
    public final Claw depositClaw;
    public final Slide extendo;
    public final Slide deposit;
    public static ArrayList<ControlledMotor> motors = new ArrayList<>();
    //public final ColorSensor color;

    /**
     * This initializes the arm and slide motors and resets the mode to the default.
     * This should be run before any other methods.
     *
     * @param hardwareMap The hardware map to use to get the motors.
     */
    public MotorControl(@NonNull HardwareMap hardwareMap) {
        extendo = new Slide(hardwareMap, // port 0 of exp hub and chub, equiv to left_back I think
                "extendo",
                new PIDFController.PIDCoefficients(3,0,0));
        extendo.setEncoder(hardwareMap.get(DcMotorEx.class,"left_back"));
        deposit = new Slide(hardwareMap, // port 1 of exp hub and chub,
                // equiv to left_front I think could be wrong though
                "deposit",
                new PIDFController.PIDCoefficients(3,0,0));
        deposit.setEncoder(hardwareMap.get(DcMotorEx.class,"right_back"));


        extendoClaw = new Claw(hardwareMap.get(Servo.class, "extendoClaw"));
        depositClaw = new Claw(hardwareMap.get(Servo.class, "depositClaw"));

        //color = hardwareMap.get(ColorSensor.class, "color");
    }

    /**
     * This class updates the arm and slide motors to match the current state.
     */
    public void update() {
        for (ControlledMotor motor : motors) {
            motor.update();
        }
    }


    public boolean closeEnough() {
        return motors.stream().allMatch(ControlledMotor::closeEnough);
    }

    public boolean isOverCurrent() {
        return motors.stream().anyMatch(ControlledMotor::isOverCurrent);
    }

    /**
     * This class controls the slide motor.
     */
    public static class Slide extends ControlledMotor {
        boolean reversed = false;
        boolean resetting = false;
        PIDFController pid;
        public DcMotorEx encoder;
        public double encoderOffset = 0;
        /**
         * This initializes the slide motor. This should be run before any other methods.
         *
         * @param hardwareMap The hardware map to use to get the motors.
         */
        public Slide(HardwareMap hardwareMap,String motorName,PIDFController.PIDCoefficients pidCoefficients) {
            super();
            pid = new PIDFController(pidCoefficients);
            motor = hardwareMap.get(DcMotorEx.class, motorName);
            encoder = motor;
            motor.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
            motor.setCurrentAlert(6, CurrentUnit.AMPS);
            if (reversed) {
                motor.setDirection(DcMotorSimple.Direction.REVERSE);
            } else {
                motor.setDirection(DcMotorSimple.Direction.FORWARD);
            }
            motor.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);

        }

        public void setEncoder(DcMotorEx motor) {
            encoder = motor;
        }

        /**
         * This stops the slide, sets the state to down, sets the target to 0, and resets the encoder.
         */
        public void reset() {
            motor.setPower(0);
            targetPosition = 0;
            setPosition(0);
        }


        public void update() {
            pid.targetPosition = targetPosition;
            if (resetting) {
                if (motor.getCurrent(CurrentUnit.AMPS) > 2.5) { //current limit end detection
                    reset();
                    resetting = false;
                }
            } else {
                if (!motor.isOverCurrent()) {
                    motor.setPower(pid.update(getPosition()));
                } else {
                    motor.setPower(0);
                }
            }
        }

        public void findZero() {
            motor.setPower(-0.5);
            resetting = true;
        }

        public double getPosition() {
            if (reversed) {
                return (encoder.getCurrentPosition() * -1) - encoderOffset;
            } else {
                return encoder.getCurrentPosition() - encoderOffset;
            }
        }

        public void setPosition(double position) {
            // some weird math happening here
            // this *should* work because we minus in getPosition but unsure
            // TODO: prob causing problems
            encoderOffset += getPosition();
        }


        public boolean closeEnough() {
            return Math.abs(getPosition() - targetPosition) < 20;
        }

    }


    public static class Claw {
        public Servo servo;
        public boolean closed = false;
        public double openPos = 0;
        public double closedPos = 1; // TODO TUNE

        public Claw(Servo servo) {
            this.servo = servo;
            open();
        }
        public Claw(Servo servo, double openPos, double closedPos) {
            this.servo = servo;
            this.openPos = openPos;
            this.closedPos = closedPos;
            open();
        }


        public void setPosition(double position) {
            servo.setPosition(position);
        }
        public double getPosition() {
            return servo.getPosition();
        }

        public void open() {
            closed = false;
            servo.setPosition(openPos);
        }
        public void close() {
            closed = true;
            servo.setPosition(closedPos);
        }
        public void toggle() {
            if (closed) {
                open();
            } else {
                close();
            }
        }

    }

    public abstract static class ControlledMotor {
        public DcMotorEx motor;
        double targetPosition;
        public double getTargetPosition() {
            return targetPosition;
        }
        public void setTargetPosition(double targetPosition) {
            this.targetPosition = targetPosition;
        }
        public abstract void update();
        public abstract void reset();
        public abstract boolean closeEnough();
        public boolean isOverCurrent() {
            return motor.isOverCurrent();
        }

        public ControlledMotor() {
            motors.add(this);
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/motor/MotorControl.kt
package org.firstinspires.ftc.teamcode.motor

import com.acmerobotics.dashboard.config.Config
import com.qualcomm.robotcore.hardware.AnalogInput
import com.qualcomm.robotcore.hardware.CRServo
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.DigitalChannel
import com.qualcomm.robotcore.hardware.HardwareMap
import com.qualcomm.robotcore.hardware.Servo
import org.firstinspires.ftc.robotcore.external.navigation.CurrentUnit
import org.firstinspires.ftc.teamcode.helpers.CachingDcMotorEx
import org.firstinspires.ftc.teamcode.helpers.Color
import org.firstinspires.ftc.teamcode.helpers.LogTelemetry
import org.firstinspires.ftc.teamcode.helpers.MotorGroup
import org.firstinspires.ftc.teamcode.helpers.RGBLight
import org.firstinspires.ftc.teamcode.helpers.ServoGroup
import org.firstinspires.ftc.teamcode.helpers.control.PIDFController
import kotlin.math.abs
import kotlin.math.sign
import kotlin.math.sqrt


@Config
class MotorControl(hardwareMap: HardwareMap, lateinit: Boolean = false) {
    @JvmField
    val extendoArm = ThreeArm(hardwareMap.get(Servo::class.java, "sArm"), 0.99, 0.89, 0.95) // DOWN UP MID

    val intake = Intake(hardwareMap.get(CRServo::class.java, "intake"))


    val extendoMotors = MotorGroup(
        hardwareMap.get(DcMotorEx::class.java, "extendo1"),
        hardwareMap.get(DcMotorEx::class.java, "extendo2")
    )

    init {
        extendoMotors.setDirections(
            DcMotorSimple.Direction.REVERSE,
            DcMotorSimple.Direction.FORWARD
        )
    }

    @JvmField
    val extendo: Slide = Slide(
        CachingDcMotorEx(extendoMotors), // port 1 of chub and exhub, encoder is left_front
        PIDFController(Constants.extendoPID),
        encoder=Encoder(hardwareMap.get(DcMotorEx::class.java, "left_front"), true),
        reversed=false
    )

    val depositArmServo = ServoGroup(hardwareMap.servo["dArm"],hardwareMap.servo["dArm2"])

    init {
        depositArmServo.setDirections(Servo.Direction.FORWARD, Servo.Direction.REVERSE)
        depositArmServo.scaleRange(0, Constants.depArmS1min, Constants.depArmS1max)
        depositArmServo.scaleRange(1, Constants.depArmS2min, Constants.depArmS2max)
    }

    val depositArmEncoderInput: AnalogInput = hardwareMap.analogInput["depositArmEncoder"]

    val depositArmEncoder = AxonEncoder { (3.3 - depositArmEncoderInput.voltage) * 1.05 }

    @JvmField
    val depositArm = DepositArm(depositArmServo, 0.4, 0.96, 0.72, depositArmEncoder)

    @JvmField
    val depositClaw = Claw(hardwareMap.get(Servo::class.java, "depositClaw"), 0.79, 0.48) // 0.35 0.1 // 0.55 0.1

    val depositMotors = MotorGroup(
        hardwareMap.get(DcMotorEx::class.java, "deposit1"),
        hardwareMap.get(DcMotorEx::class.java, "deposit2")
    )
    init {
        depositMotors.setDirections(DcMotorSimple.Direction.REVERSE, DcMotorSimple.Direction.FORWARD)
    }

    @JvmField
    val deposit: Slide = Slide(
        CachingDcMotorEx(
            depositMotors
        ),
        // port 0 of exp hub and 0 of chub,
        // encoder is left_back
        PIDFController(
            Constants.depositPID
        ), // { a, b -> 0.05 }, // Static feedforward
        encoder=Encoder(hardwareMap.get(DcMotorEx::class.java, "right_front"), true),
        reversed=false
    )

    val dColor = BLColor(hardwareMap.digitalChannel["digital0"],hardwareMap.digitalChannel["digital1"])

    val eColor = BLColor(hardwareMap.digitalChannel["digital2"],hardwareMap.digitalChannel["digital3"])


    val extendoArmEncoder = AxonEncoder(hardwareMap.analogInput["extendoArmEncoder"])

    val topLight = RGBLight(hardwareMap.servo.get("rgb"))

    val motors = listOf(extendo, deposit)

    var resetting = true

    init {
        if (!lateinit) {
            init()
        }
    }
    
    val logger = LogTelemetry("MotorControl/")

    fun init() {
        topLight.color = Color.YELLOW
        val depositArmStart = depositArmEncoder.position
        if (depositArmEncoder.position > 0.5) {
            // encoder isn't perfect
            // but this will make it not slam
            // and then the mp can deal with it
            depositArm.position = depositArmEncoder.position
        } else {
            depositArm.position = 0.4
        }
        extendoArm.moveDown()

        intake.stop()
        depositClaw.open()

        extendoArm.moveMid()

        motors.forEach { it.findZero() }
    }

    /**
     * This class updates the arm and slide motors to match the current state.
     */
    fun update() {
        motors.forEach { it.update() }

        logger.write("extendoArm/target", extendoArm.position)
        logger.write("extendoArm/actual", extendoArmEncoder.position)
        logger.write("depositArm/target", depositArm.position)
        logger.write("depositArm/actual", depositArmEncoder.position)
        logger.write("deposit/target", deposit.targetPosition)
        logger.write("deposit/actual", deposit.position)
        logger.write("extendo/target", extendo.targetPosition)
        logger.write("extendo/actual", extendo.position)
        logger.write("dColor/color", dColor.color)
        logger.write("eColor/color", eColor.color)
        logger.write("intake/power", intake.servo.power)
        logger.write("depositClaw/target", depositClaw.position)
        logger.write("topLight/colorPos", topLight.servo.position)
        logger.write("topLight/color", topLight.color)

        logger.update()


        if (resetting && motors.none { it.resetting }) {
            topLight.color = Color.GREEN
            resetting = false
        }
    }


    fun closeEnough(): Boolean {
        return motors.all { it.closeEnough() }
    }

    val isOverCurrent: Boolean
        get() = motors.any { it.isOverCurrent }

    class AxonEncoder(val getter: () -> Double) {
        constructor(pin: AnalogInput): this(pin::getVoltage)
        val position
            get() = getter() / 3.3

        val posDegrees
            get() = getter() / 3.3 * 360
    }

    class BLColor(val pin0: DigitalChannel, val pin1: DigitalChannel) {
        fun boolsToColor(first: Boolean, second: Boolean): Color {
            return if (first) {
                if (second) {
                    Color.YELLOW
                } else {
                    Color.BLUE
                }
            } else if (second) {
                Color.RED
            } else {
                Color.NONE
            }
        }
        val color: Color
            get() {
                // this could be EXTREMELY inefficient and trigger a read,
                // but as long as you remember to use bulk reads, it's fine :)
                return boolsToColor(pin0.state, pin1.state)
            }

    }

    class Encoder(val motor: DcMotor, val reversed: Boolean = false) {
        private var offset = 0.0

        var position: Double
            get() = if (!reversed) {
                (motor.currentPosition * -1) - offset
            } else {
                motor.currentPosition - offset
            }
            set(newPosition) {
                // some weird math happening here
                // this *should* work because we minus in getPosition but unsure
                // TODO: prob causing problems
                offset = motor.currentPosition - newPosition
            }

        fun reset() {
            position = 0.0
        }
    }

    class Slide(motor: DcMotorEx, val pid: PIDFController, val encoder: Encoder, val reversed: Boolean = false) :
        ControlledMotor(motor) {
        var resetting = false

        init {
            motor.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
            motor.setCurrentAlert(6.0, CurrentUnit.AMPS)
            if (reversed) {
                motor.direction = DcMotorSimple.Direction.REVERSE
            } else {
                motor.direction = DcMotorSimple.Direction.FORWARD
            }
            motor.mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER
            targetPosition = 20.0
        }

        /**
         * This stops the slide, sets the state to down, sets the target to 0, and resets the encoder.
         */
        override fun reset() {
            motor.power = 0.0
            targetPosition = 0.0
            position = 0.0
        }


        override fun update() {
            pid.targetPosition = targetPosition
            if (resetting) {
                if (motor.getCurrent(CurrentUnit.AMPS) > 2.5) { //current limit end detection
                    reset()
                    resetting = false
                }
            } else {
                if (!motor.isOverCurrent) {
                    val output = pid.update(position)
                    motor.power = sign(output) * sqrt(abs(output))
                    //motor.power = output
                } else {
                    motor.power = 0.0
                }
            }
        }

        override fun findZero() {
            motor.power = -0.5
            resetting = true
        }

        var position: Double
            get() = encoder.position
            set(position) { encoder.position = position }


        override fun closeEnough(): Boolean {
            return abs(position - targetPosition) < 25
        }
    }


    class Claw(val servo: Servo, val openPos: Double, val closedPos: Double) {
        var closed: Boolean = false

        var position: Double
            get() = servo.position
            set(position) {
                servo.position = position
            }

        fun open() {
            closed = false
            servo.position = openPos
        }

        fun close() {
            closed = true
            servo.position = closedPos
        }

        fun toggle() {
            if (closed) {
                open()
            } else {
                close()
            }
        }
    }

    open class ServoArm(val servo: Servo) {
        val down: Boolean
            get() {
                return position == downPos
            }
        var upPos: Double = 0.2
        var downPos: Double = 0.03 // TODO TUNE

        constructor(servo: Servo, downPos: Double, upPos: Double) : this(servo) {
            this.downPos = downPos
            this.upPos = upPos
        }


        var position: Double = 0.0 // cache last sent value to read from
            set(position) {
                servo.position = position
                field = position
            }

        fun moveUp() {
            position = upPos
        }

        fun moveDown() {
            position = downPos
        }

        fun toggle() {
            if (down) {
                moveUp()
            } else {
                moveDown()
            }
        }
    }

    open class ThreeArm(servo: Servo,
                   downPos: Double = 0.03, //wrong
                   upPos: Double = 0.2, // wrong
                   val midPos: Double = 0.6) // tuned
        : ServoArm(servo, downPos,upPos) {
        val mid: Boolean
            get() {
                return abs(position -  midPos) < 0.05
            }

        fun moveMid() {
            position = midPos
        }
    }

    class DepositArm(servo: Servo, downPos: Double, upPos: Double, midPos: Double, encoder: AxonEncoder): ThreeArm(servo, downPos, upPos, midPos) {
        fun moveForward() = moveUp()
        fun moveBack() = moveDown()
        fun moveHang() = moveMid()
    }


    class Intake(val servo: CRServo, val inSpeed: Double = 1.0, val outSpeed: Double = -1.0) {
        fun intake() {
            servo.power = inSpeed
        }

        fun eject() {
            servo.power = outSpeed
        }

        fun stop() {
            servo.power = 0.0
        }
    }

    abstract class ControlledMotor(val motor: DcMotorEx) {
        @JvmField
        var targetPosition = 0.0

        abstract fun update()
        abstract fun reset()
        abstract fun closeEnough(): Boolean
        val isOverCurrent: Boolean
            get() = motor.isOverCurrent

        abstract fun findZero()

    }
}
